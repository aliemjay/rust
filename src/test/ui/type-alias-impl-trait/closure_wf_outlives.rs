// If the hidden type is a closure, we used to require the "outlives" bounds
// that appear on the defining site to also appear on the opaque type.

// check-fail

#![feature(type_alias_impl_trait)]
#![allow(dead_code)]

mod test1 {
    type Opaque<'a, 'b> = impl Sized + 'a + 'b;

    fn define<'a, 'b>() -> Opaque<'a, 'b>
    where
        'a: 'b,
    {
        || {}
    }
}

// Same as the above but through indirection `'x`
mod test2 {
    type Opaque<'a, 'b> = impl Sized + 'a + 'b;

    fn define<'a, 'b, 'x>() -> Opaque<'a, 'b>
    where
        'a: 'x,
        'x: 'b,
    {
        || {}
    }
}

mod test2_fixed {
    type Opaque<'a: 'b, 'b> = impl Sized + 'a + 'b;

    fn define<'a, 'b, 'x>() -> Opaque<'a, 'b>
    where
        'a: 'x,
        'x: 'b,
    {
        || {}
    }
}

// requires `T: 'static`
mod test3 {
    type Opaque<T> = impl Sized;
    //~^ ERROR the parameter type `T` may not live long enough

    fn define<T>() -> Opaque<T>
    where
        T: 'static,
    {
        || {}
    }
}

fn main() {}
