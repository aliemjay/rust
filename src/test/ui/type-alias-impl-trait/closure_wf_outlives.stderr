error[E0310]: the parameter type `T` may not live long enough
  --> $DIR/closure_wf_outlives.rs:47:22
   |
LL |     type Opaque<T> = impl Sized;
   |                      ^^^^^^^^^^ ...so that the type `T` will meet its required lifetime bounds...
   |
note: ...that is required by this bound
  --> $DIR/closure_wf_outlives.rs:52:12
   |
LL |         T: 'static,
   |            ^^^^^^^
help: consider adding an explicit lifetime bound...
   |
LL |     type Opaque<T: 'static> = impl Sized;
   |                  +++++++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0310`.
